src/python/pants/backend/url_handlers/s3/register.py

```diff
@@ -50,44 +51,75 @@ async def access_aws_credentials() -> AWSCredentials:
     return AWSCredentials(creds)
 
 
-# NB: The URL is expected to be in path-style
-# See https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html
-def _get_aws_auth_headers(url: str, aws_credentials: AWSCredentials):
-    from botocore import auth  # pants: no-infer-dep
+@dataclass(frozen=True)
+class S3DownloadFile:
+    region: str
+    bucket: str
+    key: str
+    query: str
+    expected_digest: FileDigest
+
 
-    request = SimpleNamespace(
-        url=url,
+@rule
+async def download_from_s3(request: S3DownloadFile, aws_credentials: AWSCredentials) -> Digest:
+    from botocore import auth, exceptions  # pants: no-infer-dep
+
+    # NB: The URL for auth is expected to be in path-style
+    path_style_url = "https://s3"
+    if request.region:
+        path_style_url += f".{request.region}"
+    path_style_url += f".amazonaws.com/{request.bucket}/{request.key}"
+    if request.query:
+        path_style_url += f"?{request.query}"
+
+    http_request = SimpleNamespace(
+        url=path_style_url,
         headers={},
         method="GET",
         auth_path=None,
     )
-    auth.HmacV1Auth(aws_credentials.creds).add_auth(request)
-    return request.headers
+    # NB: The added Auth header doesn't need to be valid when accessing a public bucket. When
+    # hand-testing, you MUST test against a private bucket to ensure it works for private buckets too.
+    signer = auth.HmacV1Auth(aws_credentials.creds)
+    try:
+        signer.add_auth(http_request)
+    except exceptions.NoCredentialsError:
+        pass  # The user can still access public S3 buckets without credentials
+
+    virtual_hosted_url = f"https://{request.bucket}.s3"
+    if request.region:
+        virtual_hosted_url += f".{request.region}"
+    virtual_hosted_url += f".amazonaws.com/{request.key}"
```

Comment:
Is `request.key` guaranteed to not start with a `/`?

Reply:
Yeah. I should add an assert